source/common/upstream/upstream_impl.h

```diff
@@ -166,27 +166,51 @@ class HostImpl : public HostDescriptionImpl,
   std::atomic<bool> used_;
 };
 
-typedef std::shared_ptr<std::vector<HostSharedPtr>> HostVectorSharedPtr;
-typedef std::shared_ptr<const std::vector<HostSharedPtr>> HostVectorConstSharedPtr;
-typedef std::shared_ptr<std::vector<std::vector<HostSharedPtr>>> HostListsSharedPtr;
-typedef std::shared_ptr<const std::vector<std::vector<HostSharedPtr>>> HostListsConstSharedPtr;
+class HostsPerLocalityImpl : public HostsPerLocality {
+public:
+  HostsPerLocalityImpl() : HostsPerLocalityImpl(false) {}
+
+  HostsPerLocalityImpl(bool has_local_locality)
+      : HostsPerLocalityImpl(has_local_locality ? std::vector<HostVector>({{}})
+                                                : std::vector<HostVector>()) {}
+
+  HostsPerLocalityImpl(std::vector<HostVector>&& locality_hosts)
+      : HostsPerLocalityImpl(std::move(locality_hosts), !locality_hosts.empty()) {}
```

Comment:
It seems fragile to assume that there are local locality hosts if the vector is not empty, given the underlying reasoning behind this change. Should we remove this construction version?

Reply:
Yeah, this is there for tests, I can replace this with util.